File: CH7(Backtracking)/Number_of_Permutations.py
n, k = map(int, input().split())

p = []
used = [False] * 8
ans = 0


def solve():
    cnt = 0

    for i in range(n):
        for j in range(i + 1, n):
            if p[i] > p[j]:
                cnt += 1

    return cnt


def build():
    global ans

    if len(p) == n:
        if solve() == k:
            ans += 1
        return

    for i in range(1, n + 1):
        if not used[i]:

            p.append(i)
            used[i] = True

            build()

            used[i] = False
            p.pop()


if __name__ == "__main__":
    build()
    print(ans)

File: CH7(Backtracking)/Number_of_Permutations.cpp
#include <bits/stdc++.h>
using namespace std;

int n, k;
vector<int> p;
bool used[8];
long long ans = 0;

int solve(){
    int cnt = 0;

    for(int i = 0; i < n; i++){
        for(int j = i + 1; j < n; j++){
            if(p[i] > p[j]){
                cnt++;
            }
        }
    }

    return cnt;
}

void build(){
    if(p.size() == n){
        if(solve() == k){
            ans++;
        }
        return;
    }

    for(int i = 1; i <= n; i++){
        if(!used[i]){

            p.push_back(i);
            used[i] = true;

            build();

            used[i] = false;
            p.pop_back();
        }
    }
}

int main(){
    cin >> n >> k;

    build();

    cout << ans;
}